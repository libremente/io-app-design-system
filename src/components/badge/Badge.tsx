import React from "react";
import { Platform, StyleSheet, Text, View } from "react-native";
import {
  IOBadgeHSpacing,
  IOBadgeRadius,
  IOBadgeVSpacing,
  IOColors,
  useIOExperimentalDesign
} from "../../core";
import { makeFontStyleObject } from "../../utils/fonts";
import { WithTestID } from "../../utils/types";

export type Badge = WithTestID<{
  text: string;
  variant:
    | "default"
    | "info"
    | "warning"
    | "error"
    | "success"
    | "purple"
    | "lightBlue"
    | "blue"
    | "turquoise"
    | "contrast";
}>;

type VariantProps = {
  foreground: IOColors;
  background: IOColors;
};

const mapVariants: Record<NonNullable<Badge["variant"]>, VariantProps> = {
  default: {
    foreground: "grey-700",
    background: "grey-50"
  },
  info: {
    foreground: "info-850",
    background: "info-100"
  },
  warning: {
    foreground: "warning-850",
    background: "warning-100"
  },
  success: {
    foreground: "success-850",
    background: "success-100"
  },
  error: {
    foreground: "error-850",
    background: "error-100"
  },
  purple: {
    foreground: "hanPurple-850",
    background: "hanPurple-100"
  },
  lightBlue: {
    foreground: "blueIO-850",
    background: "blueIO-50"
  },
  blue: {
    foreground: "white",
    background: "blueIO-500"
  },
  turquoise: {
    foreground: "turquoise-850",
    background: "turquoise-50"
  },
  contrast: {
    foreground: "grey-700",
    background: "white"
  }
};

const styles = StyleSheet.create({
  badge: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
    ...Platform.select({
      android: {
        textAlignVertical: "center"
      }
    }),
    borderRadius: IOBadgeRadius,
    paddingHorizontal: IOBadgeHSpacing,
    paddingVertical: IOBadgeVSpacing
  },
  label: {
    fontSize: 12,
    lineHeight: 16,
    alignSelf: "center",
    textTransform: "uppercase",
    flexShrink: 1
  },
  labelFont: {
    ...makeFontStyleObject("Regular", false, "ReadexPro")
  },
  // TODO: Remove this when legacy look is deprecated https://pagopa.atlassian.net/browse/IOPLT-153
  legacyLabelFont: {
    ...makeFontStyleObject("SemiBold", false, "TitilliumWeb")
  }
});

/**
 * Official badge component
 */
export const Badge = ({ text, variant, testID }: Badge) => {
  const { isExperimental } = useIOExperimentalDesign();
  return (
    <View
      testID={testID}
      style={[
        styles.badge,
        { backgroundColor: IOColors[mapVariants[variant].background] }
      ]}
    >
      <Text
        numberOfLines={1}
        ellipsizeMode="tail"
        style={[
          styles.label,
          isExperimental ? styles.labelFont : styles.legacyLabelFont,
          { color: IOColors[mapVariants[variant].foreground] }
        ]}
      >
        {text}
      </Text>
    </View>
  );
};
