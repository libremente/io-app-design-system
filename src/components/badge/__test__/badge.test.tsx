import React from "react";
import * as TestRenderer from "react-test-renderer";
import { Badge } from "../Badge";

describe("Test Badge Components", () => {
  it("Badge Snapshot", () => {
    const badge = TestRenderer.create(
      <Badge text={"text"} variant={"default"}></Badge>
    ).toJSON();
    expect(badge).toMatchSnapshot();
  });
});
