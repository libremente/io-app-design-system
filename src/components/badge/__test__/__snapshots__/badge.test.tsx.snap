// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Badge Components Badge Snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 24,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 8,
        "paddingVertical": 4,
      },
      {
        "backgroundColor": "#F4F5F8",
      },
    ]
  }
>
  <Text
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      [
        {
          "alignSelf": "center",
          "flexShrink": 1,
          "fontSize": 12,
          "lineHeight": 16,
          "textTransform": "uppercase",
        },
        {
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "600",
        },
        {
          "color": "#555C70",
        },
      ]
    }
  >
    text
  </Text>
</View>
`;
